### PR TITLE
ENH/MAINT: Version switcher from the sphinx theme

### LINKED ISSUE #15377: ENH: change version switcher to use the theme
The version switcher is now available on the theme side so we can update the badge logic we added in #14132

https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/configuring.html?highlight=version#add-a-dropdown-to-switch-between-docs-versions

One thing I am not sure yet is if they handle colours like we do. If not I will raise it upstream as I think it's a handy visual help.

I believe this would close https://github.com/scipy/scipy.org/issues/423 and https://github.com/scipy/docs.scipy.org/issues/59

### PR DESCRIPTION
**Pull request by tupui**
Closes #15377

Use the version switcher provided by the theme instead of our custom CSS/HTML solution.

I added all versions I saw on https://docs.scipy.org. I am still unsure about how to handle the dev and PR versions. Let see how it goes.

Once this is done, we can decide if we want to remove https://docs.scipy.org or not. The caveat of this method is that when switching to an old version, the version switcher will disappear. So we have to go back to latest to change again the switcher if we want to go to another version.

### CODE REVIEW

**Comment by tupui** [comment]
@jorisvandenbossche is this really released (it's live in the doc although there is no doc for 0.7.2, just 0.7.1)? On master I can see the template but not on the release branch 0.7.x.

Also I have a question, for SciPy we have a dynamic url when we make a PR. How should I handle this case? For the dev deployed on our main vs released versions, I think I can use some conditions in the config file as I did now. I could do something a bit more involved with re-writing the json file. But maybe there is another simpler way to define such things?

Last question I have for now, is there an automatic way to style the version switcher? If you look at our current switcher, it would change colour depending on whether it's a release, dev or something too old (and we picked sensible colours for colour deficient readers).

**Comment by jorisvandenbossche** [comment]
No, it's not yet released. Hopefully there will be a 0.8.0 shortly that includes this, until then you will have to test with the master branch.

> Also I have a question, for SciPy we have a dynamic url when we make a PR. How should I handle this case? 

Good question, I am not fully sure what the best solution is. Are you using readthedocs or something like circle ci to host the docs from the PR? 
We currently have the same problem at https://github.com/pydata/pydata-sphinx-theme itself for PRs, where the version switcher doesn't work. We should probably open an issue there. (potentially by using a relative path for the switcher json, but need to re-check some discussion related to that in https://github.com/pydata/pydata-sphinx-theme/pull/436)

> Last question I have for now, is there an automatic way to style the version switcher?

The version switcher block has a specific ids on the different elements that can be used for styling. But that's of course for general styling, identical across all versions (unless you would vary the css depending on the version).
I know that mne-python did something similar before, and also mentioned they lost this version-specific coloring while migrating to the upstream version switcher (https://github.com/mne-tools/mne-python/pull/9916, cc @drammock). So it it is something several projects would want, we should see how we can enable this (or document how to do it).

**Comment by tupui** [comment]
> No, it's not yet released. Hopefully there will be a 0.8.0 shortly that includes this, until then you will have to test with the master branch.
> 

Ok thanks 🙏 I will play with master then.

> Good question, I am not fully sure what the best solution is. Are you using readthedocs or something like circle ci to host the docs from the PR? 

It's CircleCI for us and we have a link in the PR to see the artifact.

> We currently have the same problem at https://github.com/pydata/pydata-sphinx-theme itself for PRs, where the version switcher doesn't work. We should probably open an issue there. (potentially by using a relative path for the switcher json, but need to re-check some discussion related to that in https://github.com/pydata/pydata-sphinx-theme/pull/436)


I thought I could manually rewrite the config by intercepting the final URL. A simpler way would be that if you select a version that does not exists it gives the current build. This way we don't need to know the dynamic url. That would solve everything for us I think 🤔 Because the deployed dev can have a fix "dev" version since the url is constant (what I intended to do in this PR) and for the PRs it would be good.

> The version switcher block has a specific ids on the different elements that can be used for styling. But that's of course for general styling, identical across all versions (unless you would vary the css depending on the version).
> 
> I know that mne-python did something similar before, and also mentioned they lost this version-specific coloring while migrating to the upstream version switcher (https://github.com/mne-tools/mne-python/pull/9916, cc @drammock). So it it is something several projects would want, we should see how we can enable this (or document how to do it).

If there is an id we can hack something as it's currently the case (although the release of our doc does not have the version for some reason, another problem...).

But yes having an optional flag in the json could be handy. We can change the name already, having other options could.

**Comment by drammock** [comment]
> So it it is something several projects would want, we should see how we can enable this (or document how to do it).

something that I think @larsoner suggested was that you could add a `extra_classes` field into the version JSON. For example, we might

1. add `extra_classes='dev'` here:

https://github.com/mne-tools/mne-python/blob/2d5d1686e9b06a1e7bc29794bd2084b0336f0ded/doc/_static/versions.json#L2-L5

2. the theme would dynamically add that classname to the switcher button at load time, and
3. our CSS could target that classname to override the default button styling.

Step 2 requires some new JS in the theme, the rest is done per-project.

**Comment by tupui** [comment]
> > So it it is something several projects would want, we should see how we can enable this (or document how to do it).
> 
> something that I think @larsoner suggested was that you could add a `extra_classes` field into the version JSON. 

This would be good I think.

**Comment by tupui** [comment]
0.8 just got released, I will have a look again.

**Comment by jorisvandenbossche** [comment]
I opened an issue at https://github.com/pydata/pydata-sphinx-theme/issues/566 for the version-specific styling

**Comment by tupui** [comment]
Thanks @jorisvandenbossche for opening the issue!

Can you have a look at this PR? I am not sure if I did it right. It seems to be working like that but I am not sure I understand why because it does select `"dev"` but does not redirect to `http://scipy.github.io/devdocs/` as I would have thought.

As for CircleCI, once the above is sorted out, I would just have detected that we are in the CI and disabled the version switcher.

**Comment by jorisvandenbossche** [comment]
If you check the browser console, you can see something like *"Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://github.com/tupui/scipy/blob/version_switcher/doc/source/_static/version_switcher.json. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200"*. 
We ran into that on the PR that developed the version switcher as well, and I suppose this is the reason it is not working properly.

**Comment by tupui** [comment]
Thanks @jorisvandenbossche. After disabling CORS it was not working first because the JSON was not loading properly due to an extra comma at the end of the JSON. Maybe some warning would be nice here in case it fails to load the file. That would have helped me.

```
open -n -a /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome  --args --user-data-dir=”/tmp/chrome_dev_test” --disable-web-security
```

Anyway, it's working now: when it loads, the current build is live (we are not directly redirected to `url_template` which is convenient). I can select another version and then it would follow the template url. The only issue I see now is that it cannot go back to the `dev` version because it would not match any URL (`https://docs.scipy.org/doc/scipy-dev/reference/`), but this can be fixed with a redirection on our side.

**Comment by tupui** [comment]
I've fixed the path for this to be merged. To see how it behaves, see the previous CI here
https://41824-1460385-gh.circle-artifacts.com/0/html-scipyorg/index.html

File: doc/source/_static/version_switcher.json [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    },
+    {
+        "version": "1.7.2"
+    },
+    {
+        "version": "1.7.1"
+    },
+    {
+        "version": "1.7.0"
+    },
```
Comment (tupui) [inline doc/source/_static/version_switcher.json:22]: These will depends on how we fix the deployment of the live docs. Previous versions' links are resolving properly.

**Comment by tupui** [comment]
> Yay! It is a nice feature. Maybe a subsequent PR can consider an optimal color that will blend in with the rest of the site.

Yes totally! Joris opened https://github.com/pydata/pydata-sphinx-theme/issues/566 to do this on the theme side. I would want to set the colours like I did before this (some red, orange, green which are colour deficient friendly). If we don't want to wait, we could tweak the CSS to do it.

**Comment by jorisvandenbossche** [comment]
> After disabling CORS

Wondering, how did you do that? (I know we struggled with this while developing the feature in the upstream theme)

>  the JSON was not loading properly due to an extra comma at the end of the JSON. Maybe some warning would be nice here in case it fails to load the file

That could be a browser console warning? 

>  Maybe a subsequent PR can consider an optimal color that will blend in with the rest of the site.

To be clear, the issue I opened is to have version-dependent colors, but we already added css classes to the dropdown button so you can robustly give it another color (one color for all versions).

**Comment by tupui** [comment]
> Wondering, how did you do that? (I know we struggled with this while developing the feature in the upstream theme)

On Safari you have an option in the `Develop` menu. And for Chrome you can launch it from the terminal with option (be sure to close all sessions or it will just bring the current one up and not disable anything)

```
open -n -a /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome  --args --user-data-dir=”/tmp/chrome_dev_test” --disable-web-security
```

> That could be a browser console warning?

Yes would be helpful I think.

> ... (one color for all versions).

I did not think about this point. It might look like a rainbow tree if we allow different colour for different version when you open the list. Sections might be better and maybe then just having the proper colour only for the selected version.

**Comment by jorisvandenbossche** [comment]
Thanks for the info on CORS disabling, that's useful :)

> >  ... (one color for all versions).
>
> I did not think about this point. It might look like a rainbow tree if we allow different colour for different version when you open the list. Sections might be better and maybe then just having the proper colour only for the selected version.

Ah, I was actually talking about the color of the "button", not of the entries in the dropdown list. But maybe I misunderstood the original feature request about this. 
I agree sections in the dropdown would be nice, but that's probably not straightforward with the current way we work with the json file to list the entries ..

**Comment by tupui** [comment]
> Ah, I was actually talking about the color of the "button", not of the entries in the dropdown list. But maybe I misunderstood the original feature request about this. I agree sections in the dropdown would be nice, but that's probably not straightforward with the current way we work with the json file to list the entries ..

You got it correctly the first time, it's me who misunderstood your last message 😅 But we are on the same page. No rainbow and yes section are nice but not for now.

**Comment by tupui** [comment]
My PR for the extra URL got merged on the one about extra classes will be soon. So if it's not urgent we can wait for a release of the theme and I can update this PR. Let me know what you think.

cc @tylerjereddy @rgommers

**Comment by rgommers** [comment]
That sounds fine to me, thanks @tupui

**Comment by tupui** [comment]
The theme just got released, I will update the PR next week 😃

**Comment by tupui** [comment]
@rgommers I updated the PR. We still need to decide on the style. There are 2 customisation options: style the button, style the elements in the list.

Right now it renders as:
Screenshot 2022-03-28 at 14 19 01

Button style is by default Blue and depending on if it's a dev, stable or old version the background is adjusted in the list (I made a typo in the css this is why dev is not orange).

I am not sure about colouring the list. It's a bit too rainbow-ish. Maybe just styling the button when a version is selected would be enough.

What do you think?

**Comment by tupui** [comment]
Note: before merging apply
```diff
- "json_url": "https://raw.githubusercontent.com/tupui/scipy/version_switcher/doc/source/_static/version_switcher.json",
+ "json_url": "https://scipy.github.io/devdocs/_static/version_switcher.json"
```

**Comment by tupui** [comment]
Ok I have it for the button, here are the results:
Screenshot 2022-03-28 at 17 04 01
Screenshot 2022-03-28 at 17 01 58
Screenshot 2022-03-28 at 17 30 56
Screenshot 2022-03-28 at 17 39 55

**Comment by rgommers** [comment]
That color scheme looks good to me, at least for now. It can always be tweaked later.

File: doc/source/conf.py [hunk 2]
```diff
@@ -189,9 +188,17 @@
 html_theme_options = {
   "logo_link": "index",
   "github_url": "https://github.com/scipy/scipy",
-  "navbar_start": ["navbar-logo", "version"],
+  "navbar_end": ["version-switcher", "navbar-icon-links"],
+  "switcher": {
+      "json_url": "https://raw.githubusercontent.com/tupui/scipy/version_switcher/doc/source/_static/version_switcher.json",
+      # "json_url": "https://scipy.github.io/devdocs/_static/version_switcher.json"
```
Comment (tupui) [inline doc/source/conf.py:194]:
```suggestion
      "json_url": "json_url": "https://scipy.github.io/devdocs/_static/version_switcher.json"
```

**Comment by tupui** [comment]
> That color scheme looks good to me, at least for now. It can always be tweaked later.

Ok then ready on my side @rgommers @tylerjereddy

**Comment by tupui** [comment]
I plan to merge it on Saturday night if there are no comment since then.

**Pull request merged**